project/src/com/google/daggerquery/executor/dagger_query.bzl

```diff
@@ -0,0 +1,30 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+load("@rules_java//java:defs.bzl", "java_binary")
+load("//src/com/google/daggerquery/plugin:dagger_query_textproto.bzl", "dagger_query_textproto")
+
+def dagger_query(name, dagger_app_target):
+   dagger_query_textproto(
+       name = "binding_graph_data",
+       dagger_app_target = dagger_app_target
+   )
+
+   java_binary(
+     name = name,
+     srcs = ["//src/com/google/daggerquery/executor:QueryExecutor.java"],
```

Comment:
I've tried this:
```
def dagger_query(name, dagger_app_target):
   dagger_query_textproto(
       name = "binding_graph_data",
       dagger_app_target = dagger_app_target
   )

   java_binary(
     name = name,
     runtime_deps = ["//src/com/google/daggerquery/executor:query_executor"],
     resources = [":binding_graph_data"],
   )
```
But without specifying `main_class` it complains about it with `ClassNotFoundException`: **Could not find or load main class com.google.daggerquery.dagger_query_app**. 

I can pass main class `main_class = "com.google.daggerquery.executor.QueryExecutor"` and get `ClassNotFoundException` caused by not being able to find the Query class (because `runtime_deps` is not enough and we need these dependencies at compile time).

Please, check new commit now, I've tried another approach and removed `java_binary` from executor/BUILD.

Reply:
Ah, my bad, I looked at https://docs.bazel.build/versions/master/be/java.html#java_binary again and there's a snippet at the bottom:

> The following code snippet illustrates a common mistake:
```
 java_binary(
    name = "DontDoThis",
    srcs = [
        ...,
        "GeneratedJavaFile.java",  # a generated .java file
    ],
    deps = [":generating_rule",],  # rule that generates that file
)
```
> Do this instead:
```
java_binary(
    name = "DoThisInstead",
    srcs = [
        ...,
        ":generating_rule",
    ],
)
```

so I wonder if we could do:

```
java_binary(
  name  = name,
  srcs = ["//src/com/google/daggerquery/executor:query_executor"],
  resources = [":binding_graph_data"],
)
```
and keep the original `:query_executor` rule?

Apologies for the inconvenience, if it still doesn't work let's go with your solution for now :smiley:

Reply:
It fails again with this reason:  `'//src/com/google/daggerquery/executor:query_executor' does not produce any java_binary srcs files (expected .java, .srcjar or .properties). Since this rule was created by the macro 'dagger_query', the error might have been caused by the macro implementation.`

It's strange, I thought that `java_binary` is a generating rule 🤔

Reply:
Ah you might need to change `//src/com/google/daggerquery/executor:query_executor` to `java_library` because it becomes just another target depended from here. Anyway let's not worry too much about this, we can revisit this later if we have time :smiley: